core/src/block.rs

```diff
@@ -989,6 +990,177 @@ impl From<&CommittedBlock> for Vec<Event> {
+    }
+
+    /// Revalidate transaction hashes and produce [`VersionedCommittedBlock`]
+    ///
+    /// # Errors
+    /// - If transaction hashes don't match the hashes stored in the block header
+    pub fn revalidate_hashes(self) -> Result<VersionedCommittedBlock, eyre::Report> {
+        self.into_v1()
+            .revalidate_hashes()
+            .map(VersionedCommittedBlock::from)
+    }
+}
+
+/// Block state used to transfer accepted by consensus block through network to the other peers.
+/// This block state is not entirely trusted and require hash revalidation to obtain `CommittedBlock`.
+#[version_with_scale(n = 1, versioned = "VersionedCandidateCommittedBlock")]
+#[derive(Debug, Clone, Decode, Encode, IntoSchema, Serialize)]
+pub struct CandidateCommittedBlock {
+    /// Block header
+    pub header: BlockHeader,
+    /// Array of rejected transactions.
+    pub rejected_transactions: Vec<VersionedRejectedTransaction>,
+    /// Array of transactions, which successfully passed validation and consensus step.
+    pub transactions: Vec<VersionedValidTransaction>,
+    /// Event recommendations.
+    pub event_recommendations: Vec<Event>,
+    /// Signatures of peers which approved this block
+    pub signatures: SignaturesOf<Self>,
+}
+
+impl CandidateCommittedBlock {
+    /// Calculate the hash of the current block.
+    /// `CommitedBlock` should have the same hash as `ValidBlock`.
+    #[inline]
+    pub fn hash(&self) -> HashOf<Self> {
+        HashOf::new(&self.header).transmute()
+    }
+
+    /// Return the signatures (as `payload`) that are verified with the `hash` of this block.
+    #[inline]
+    pub fn verified_signatures(&self) -> impl Iterator<Item = &SignatureOf<Self>> {
+        self.signatures.verified_by_hash(self.hash())
+    }
+
+    /// Revalidate transaction hashes, verify signatures and produce [`CommittedBlock`]
+    ///
+    /// # Errors
+    /// - If transaction hashes don't match the hashes stored in the block header
+    /// - If signatures verification fails
+    pub fn revalidate(self, topology: &Topology) -> Result<CommittedBlock, eyre::Report> {
+        topology
+            .filter_signatures_by_roles(
+                &[
+                    Role::ValidatingPeer,
+                    Role::Leader,
+                    Role::ProxyTail,
+                    Role::ObservingPeer,
+                ],
+                self.verified_signatures(),
+            )
+            .len()
+            .ge(&topology.min_votes_for_commit())
+            .then_some(())
+            .ok_or_else(|| {
+                eyre!("The block doesn't have enough valid signatures to be committed.")
+            })?;
+        self.revalidate_hashes()
+    }
+
+    /// Revalidate transaction hashes and produce [`CommittedBlock`]
+    ///
+    /// # Errors
+    /// - If transaction hashes don't match the hashes stored in the block header
+    pub fn revalidate_hashes(self) -> Result<CommittedBlock, eyre::Report> {
```

Comment:
I would also add a comment about favoring the use of `revalidate` and that the signatures are expected to be validated before calling this function